### PR TITLE
feat: Blobstream errors

### PR DESCRIPTION
**Pull request by JordyRo1**
- [x] issue #26 
- [x] follows contribution [guide](../CONTRIBUTING.md)
- [x] code change includes tests
- [x] breaking change

Contract errors and implementation of the function using it

### CODE REVIEW

**Comment by JordyRo1** [comment]
If necessary I can also finish everything related to the general functions implementation (tests + misssing functions), but I don't know really if I should be doing it here.

File: src/lib.cairo [hunk 1]
```diff
@@ -67,6 +107,182 @@ mod Blobstream {
...
+        ///
+        /// The validator set hash that is signed over is domain separated as per
+        /// `domain_separate_validator_set_hash`.
+        /// # Arguments
+        /// `_new_nonce` - The new event nonce.
+        /// `_old_nonce` - The nonce of the latest update to the validator set.
+        /// `_new_power_threshold` - At least this much power must have signed.
+        /// `_new_validator_set_hash` - The hash of the new validator set.
+        /// `_current_validator_set`- The current validator set.
+        /// `_sigs` - Signatures
```
Comment (drspacemn) [inline src/lib.cairo:126]:
```suggestion
        /// * `_new_nonce` - The new event nonce.
        /// * `_old_nonce` - The nonce of the latest update to the validator set.
        /// * `_new_power_threshold` - At least this much power must have signed.
        /// * `_new_validator_set_hash` - The hash of the new validator set.
        /// * `_current_validator_set`- The current validator set.
        /// * `_sigs` - Signatures
```

Let's go with `*` before the args, looks to be the way the compiler comments

File: src/lib.cairo [hunk 2]
```diff
@@ -67,6 +107,182 @@ mod Blobstream {
...
+                _new_nonce, _new_power_threshold, _new_validator_set_hash
+            );
+            check_validator_signatures(
+                @self, _current_validator_set, _sigs, new_checkpoint, current_power_threshold
+            );
+
+            self.state_last_validator_checkpoint.write(new_checkpoint);
+            self.state_power_threshold.write(_new_power_threshold);
+            self.state_event_nonce.write(_new_nonce);
+        // TODO Add event emission: ISSUE:  #25
```
Comment (drspacemn) [inline src/lib.cairo:170]: ident

File: src/lib.cairo [hunk 3]
```diff
@@ -67,6 +107,182 @@ mod Blobstream {
...
+
+            // Check that the supplied current validator set matches the saved checkpoint.
+            let current_validator_set_hash: u256 = compute_validator_set_hash(
+                _current_validator_set
+            );
+
+            // Check that the supplied current validator set matches the saved checkpoint.
+            let current_validator_set_hash: u256 = compute_validator_set_hash(
+                _current_validator_set
+            ); // TODO: compute_validator_set_hash
```
Comment (drspacemn) [inline src/lib.cairo:200]: If this currently has an issue please comment the # if not please create and issue and comment

File: src/lib.cairo [hunk 4]
```diff
@@ -67,6 +107,182 @@ mod Blobstream {
...
+                domain_separate_validator_set_hash_val == last_validator_set_checkpoint,
+                Errors::SUPPLIED_VALIDATOR_SET_INVALID
+            );
+
+            let c: u256 = domain_separate_data_root_tuple_root(_new_nonce, _data_root_tuple_root);
+            check_validator_signatures(
+                @self, _current_validator_set, _sigs, c, current_power_threshold
+            );
+
+            // EFFECTS
```
Comment (drspacemn) [inline src/lib.cairo:214]: can remove

File: src/lib.cairo [hunk 5]
```diff
@@ -67,6 +107,182 @@ mod Blobstream {
...
+
+            self.state_event_nonce.write(_new_nonce);
+            self.state_data_root_tuple_roots.write(_new_nonce, _data_root_tuple_root);
+        // TODO Add event emission: ISSUE:  #25
```
Comment (drspacemn) [inline src/lib.cairo:218]: ident.

File: src/lib.cairo [hunk 6]
```diff
@@ -28,27 +28,67 @@ struct DataRoot {
...
+    //Malformed current validator set.
+    const MALFORMED_CURRENT_VALIDATOR_SET: felt252 = 'Malformed current validator set';
+    //Validator signature does not match.
+    const INVALID_SIGNATURE: felt252 = 'Invalid signature';
+    // Submitted validator set signatures do not have enough power.
+    const INSUFFICIENT_VOTING_POWER: felt252 = 'Sub validator inadequate power';
+    // New validator set nonce must be greater than the current nonce.
+    const INVALID_VALIDATOR_SET_NONCE: felt252 = 'New set nonce > current nonce';
+    // Supplied current validators and powers do not match checkpoint.
+    const SUPPLIED_VALIDATOR_SET_INVALID: felt252 = 'Validator&power not matching cp';
```
Comment (drspacemn) [inline src/lib.cairo:41]:
```suggestion
    const SUPPLIED_VALIDATOR_SET_INVALID: felt252 = 'Validator & power not matching cp';
```

File: src/lib.cairo [hunk 7]
```diff
@@ -67,6 +107,182 @@ mod Blobstream {
...
+            self.state_data_root_tuple_roots.write(_new_nonce, _data_root_tuple_root);
+        // TODO Add event emission: ISSUE:  #25
+
+        }
+    }
+
+    /// Checks that enough voting power signed over a digest.
+    /// It expects the signatures to be in the same order as the _currentValidators.
+    /// # Arguments
+    /// `_currentValidators` - The current validators.
```
Comment (drspacemn) [inline src/lib.cairo:226]: Same `*` syntax

File: src/lib.cairo [hunk 8]
```diff
@@ -67,6 +107,182 @@ mod Blobstream {
...
+    /// * A boolean
+    fn is_sig_nil(_sig: Signature) -> bool {
+        return (_sig.r == 0 && _sig.s == 0 && _sig.y_parity == true);
+    }
+
+    /// Computes the hash of a validator set.
+    /// # Arguments
+    /// * `_validators - The validator set to hash
+    /// # Returns
+    /// * * `hash` - The result of the hashing process
```
Comment (drspacemn) [inline src/lib.cairo:273]: Can remove, we'll have the implementor add the comment

**Review by drspacemn — CHANGES_REQUESTED** [review]

**Comment by drspacemn** [comment]
> If necessary I can also finish everything related to the general functions implementation (tests + misssing functions), but I don't know if I should be doing it here.

I'll open a new one and assign to you.

File: src/lib.cairo (hunk 6 above)
Comment (JordyRo1) [reply src/lib.cairo:41]: The issue here is that if I add spaces here, I will get a size 33 and the max is 31.

Comment (drspacemn) [reply src/lib.cairo:41]: Roger, the way you have it is all good

**Comment by drspacemn** [comment]
That's weird I wonder why the formatter doesn't like that comment indent. You can keep the left formatting and we will clean up in subsequent work.

**Review by drspacemn — APPROVED** [review]
lgtm

**Comment by drspacemn** [comment]
https://github.com/all-contributors please add @JordyRo1 for code

**Comment by drspacemn** [comment]
@all-contributors please add @JordyRo1 for code

**Comment by drspacemn** [comment]
@all-contributors add @JordyRo1 code

**Comment by drspacemn** [comment]
@all-contributors add @JordyRo1 for code

**Comment by drspacemn** [comment]
@all-contributors add JordyRo1 for code

**Pull request merged**